### DISCUSSION TITLE
Welcome to siglib Discussions!

### DISCUSSION THREAD

**Discussion by johned0** [Numerix-DSP/siglib#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by JasleenKaur2308** [comment]
I am successfully able to get the results for 2-FSK using siglib as designed in the link below.
Link - https://github.com/Numerix-DSP/siglib/blob/main/Examples/CExamples/fsk.c

I wanted to design for 4-FSK using the siglib in the same way , how should i designed it and what all changes are required to get the results for 4-FSK using Siglib .

**Reply by johned0** [reply]
Hi There,
You would need to rewrite SDA_FskModulate and SDA_FskDemodulate functions to support 4-FSK.
For the modulator you just need to take 2 bits at a time and duplicate the frequencies. For the demodulator, you just need to take the existing code and duplicate the filters and modify the detector.
Best regards,
John

**Reply by JasleenKaur2308** [reply]
Thank you so much for your valuable response , and the above code is for real numbers , how can i make it work in case for complex numbers?

**Comment by johned0** [comment]
I assume you are referring to complex inputs to the modulator, in which case, you need to split the complex numbers into the appropriate bit patterns to drive the modulator.
Best, John

**Reply by JasleenKaur2308** [reply]
I am referring to the complex modulated output  , i want complex modulated output and then i need to demodulate that complex output .

**Reply by johned0** [reply]
Ah, that would be Quadrature modulation, not FSK, which you originally asked about.
Please take a look at all of the functions in mod_d.c.
Best, John

**Reply by JasleenKaur2308** [reply]
Thank you so much for your reply !

**Comment by JasleenKaur2308** [comment]
In this example - https://github.com/Numerix-DSP/siglib/blob/main/Examples/CExamples/fsk.c

It is designed for real numbers ,how can i design it for complex numbers ?

**Reply by johned0** [reply]
You need to choose the digital modulation mode you require (bpsk, qpsk, qam etc.) then use the appropriate example.
Best,
John

**Reply by JasleenKaur2308** [reply]
Hey John , i want it in 2- FSK modulation mode ..

**Reply by johned0** [reply]
2-FSK is exactly what is implemented in this example - https://github.com/Numerix-DSP/siglib/blob/main/Examples/CExamples/fsk.c
Best, John

**Reply by JasleenKaur2308** [reply]
Hi John
output of SDA_FskModulateByte is pData, which is SLData_t type.
but i want output as SLComplexRect_s,
 
my device (Bladerf) takes only complex input.
 
Thanks 
Jasleen Kaur

**Reply by johned0** [reply]
I have not used this device but you could try setting the imaginary part to zero.
J

**Comment by JasleenKaur2308** [comment]
I have gone through this whole code - https://github.com/Numerix-DSP/siglib/blob/main/Examples/CExamples/fsk.c 
How can i demodulate this modulated fsk signal output and what all the filters are required ?

**Comment by johned0** [comment]
The demodulate functions are included in the example you reference.
Filters are designed using the SIF_FirBandPassFilter () function.
You need to configure this example to match your required frequencies and bandwidths etc.
